util/ranger/detacher.go

```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+		if f, ok := cond.(*expression.ScalarFunction); ok {
+			switch f.FuncName.L {
+			case ast.EQ:
+				return NeedAddColumn4EqCond(cols, accessCond, columnValues)
+			case ast.In:
+				return NeedAddColumn4InCond(cols, accessCond, f)
+			}
+		}
+	}
+
+	return false
+}
+
+// NeedAddColumn4EqCond `tidb_shard(x) = xxx`
+// For param explanation, please refer to the function `NeedAddGcColumn4ShardIndex`.
+// It checks whether EQ condtions need to be added tidb_shard() prefix.
+// (1) columns in accessCond are all columns of the index except the first.
+// (2) every column in accessCond has a constan value
+func NeedAddColumn4EqCond(cols []*expression.Column,
+	accessCond []expression.Expression, columnValues []*valueInfo) bool {
+	valCnt := 0
+	matchedKeyFldCnt := 0
+
+	// the columns of shard index shoude be more than 2, like (tidb_shard(a),a,...)
+	if len(cols) < 2 ||
+		len(accessCond) < 2 ||
+		len(columnValues) < 2 {
+		return false
+	}
+
+	for _, cond := range accessCond[1:] {
+		if cond == nil {
+			break
+		}
+
+		f, ok := cond.(*expression.ScalarFunction)
+		if !ok || f.FuncName.L != ast.EQ {
+			return false
+		}
+
+		matchedKeyFldCnt++
+	}
+	for _, val := range columnValues[1:] {
+		if val == nil {
+			break
+		}
+		valCnt++
+	}
+
+	if matchedKeyFldCnt != len(cols)-1 ||
+		valCnt != len(cols)-1 ||
+		accessCond[0] != nil ||
+		columnValues[0] != nil {
+		return false
+	}
+
+	return true
+}
+
+// NeedAddColumn4InCond `tidb_shard(x) = xxx`
+// For param explanation, please refer to the function `NeedAddGcColumn4ShardIndex`.
+// It checks whether "IN" condtions need to be added tidb_shard() prefix.
+// (1) columns in accessCond are all columns of the index except the first.
+// (2) the first param of "IN" function shoulde be a column not a expression like `a + b`
+// (3) the rest params of "IN" function all shoulde be constant
+// (4) the first param of "IN" function shoulde be the column in the expression of first index field.
+//     e.g. uk(tidb_shard(a), a). If the condtions is `WHERE b in (1, 2, 3)`, the first param of "IN" function
+//     is `b` that's not the column in `tidb_shard(a)`.
+// @param  sf	"IN" function, e.g. `a IN (1, 2, 3)`
+func NeedAddColumn4InCond(cols []*expression.Column, accessCond []expression.Expression, sf *expression.ScalarFunction) bool {
+	if len(cols) == 0 || len(accessCond) == 0 || sf == nil {
+		return false
+	}
+
+	if accessCond[0] != nil {
+		return false
+	}
+
+	fields := ExtractColumnsFromExpr(cols[0].VirtualExpr.(*expression.ScalarFunction))
+
+	c, ok := sf.GetArgs()[0].(*expression.Column)
+	if !ok {
+		return false
+	}
+
+	for _, arg := range sf.GetArgs()[1:] {
+		if _, ok := arg.(*expression.Constant); !ok {
+			return false
+		}
+	}
+
+	if len(fields) != 1 ||
+		!fields[0].Equal(nil, c) {
+		return false
+	}
+
+	return true
+}
+
+// ExtractColumnsFromExpr get all fields from input expression virtaulExpr
+func ExtractColumnsFromExpr(virtaulExpr *expression.ScalarFunction) []*expression.Column {
```

Comment:
There is `expression.ExtractColumns`.

Reply:
I need to remove the duplicate columns in the `virtaulExpr`